tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -387,210 +389,231 @@ ParsedSemanticDefineList hlsl::CollectSemanticDefinesParsedByCompiler(CompilerIn
-    CComPtr<IDxcBlob> pErrorBlob;
-    IFT(pOutputStream->QueryInterface(&pErrorBlob));
-    outStream.flush();
-    IFT(DxcResult::Create(E_INVALIDARG, DXC_OUT_NONE, {
-        DxcOutputObject::ErrorOutput(opts.DefaultTextCodePage,
-          (LPCSTR)pErrorBlob->GetBufferPointer(), pErrorBlob->GetBufferSize())
-      }, ppResult));
-    return S_OK;
+  if (pDefines) {
+    std::string newDefines = compiler.getPreprocessor().getPredefines();
+    newDefines += pDefines;
+    compiler.getPreprocessor().setPredefines(newDefines);
   }
-  return S_OK;
-}
 
-static
-bool HasUniformParams(FunctionDecl *FD) {
-  for (auto PD : FD->params()) {
-    if (PD->hasAttr<HLSLUniformAttr>())
-      return true;
+  compiler.createASTContext();
+  compiler.setASTConsumer(std::unique_ptr<ASTConsumer>(new SemaConsumer()));
+  compiler.createSema(TU_Complete, nullptr);
+
+  const FileEntry *mainFileEntry =
+      compiler.getFileManager().getFile(StringRef(pMainFile));
+  if (mainFileEntry == nullptr) {
+    throw ::hlsl::Exception(HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));
   }
-  return false;
+  compiler.getSourceManager().setMainFileID(
+      compiler.getSourceManager().createFileID(mainFileEntry, SourceLocation(),
+                                               SrcMgr::C_User));
 }
 
-static
-void WriteUniformParamsAsGlobals(FunctionDecl *FD,
-                                 raw_ostream &o,
-                                 PrintingPolicy &p) {
-  // Extract resources first, to avoid placing in cbuffer _Params
-  for (auto PD : FD->params()) {
-    if (PD->hasAttr<HLSLUniformAttr>() &&
-        hlsl::IsHLSLResourceType(PD->getType())) {
-      PD->print(o, p);
-      o << ";\n";
-    }
-  }
-  // Extract any non-resource uniforms into cbuffer _Params
-  bool startedParams = false;
-  for (auto PD : FD->params()) {
-    if (PD->hasAttr<HLSLUniformAttr>() &&
-        !hlsl::IsHLSLResourceType(PD->getType())) {
-      if (!startedParams) {
-        o << "cbuffer _Params {\n";
-        startedParams = true;
-      }
-      PD->print(o, p);
-      o << ";\n";
+
+void SetupCompilerForPreprocess(
+    CompilerInstance &compiler, _In_ DxcLangExtensionsHelper *helper,
+    _In_ LPCSTR pMainFile, _In_ TextDiagnosticPrinter *diagPrinter,
+    _In_opt_ ASTUnit::RemappedFile *rewrite, _In_ hlsl::options::DxcOpts &opts,
+    _In_ DxcDefine *pDefines, _In_ UINT32 defineCount,
+    _In_opt_ dxcutil::DxcArgsFileSystem *msfPtr) {
+
+  SetupCompilerCommon(compiler, helper, pMainFile, diagPrinter, rewrite, opts,
+                      msfPtr);
+
+  if (pDefines) {
+    PreprocessorOptions &PPOpts = compiler.getPreprocessorOpts();
+    for (size_t i = 0; i < defineCount; ++i) {
+      CW2A utf8Name(pDefines[i].Name, CP_UTF8);
+      CW2A utf8Value(pDefines[i].Value, CP_UTF8);
+      std::string val(utf8Name.m_psz);
+      val += "=";
+      val += (pDefines[i].Value) ? utf8Value.m_psz : "1";
+      PPOpts.addMacroDef(val);
     }
   }
-  if (startedParams) {
-    o << "}\n";
+}
```

Comment:
I like how these three `SetupCompiler*` functions came out. easy to read and maintain. 👍